libs/langchain-core/src/_standard/message.ts

```diff
@@ -0,0 +1,1119 @@
+ * //   name: "calculator";
+ * //   args: {operation: string, numbers: number[]};
+ * //   id?: string;
+ * // }
+ * ```
+ */
+export type $MessageToolCallBlock<TStructure extends $MessageStructure> =
+  TStructure["tools"] extends $MessageToolSet
+    ? {
+        [K in keyof TStructure["tools"]]: {
+          readonly type: "tool_call";
+          name: K;
+          args: TStructure["tools"][K] extends $MessageToolDefinition
+            ? TStructure["tools"][K]["input"]
+            : never;
+        };
+      }[keyof TStructure["tools"]]
+    : never;
+
+/**
+ * Core interface that defines the structure of messages.
+ * This interface acts as the base for describing the various shapes of messages.
+ *
+ * @example
+ * ```ts
+ * // Basic message structure with just content blocks
+ * interface SimpleMessageStructure extends $MessageStructure {
+ *   content: {
+ *     human: ContentBlock.Text;
+ *     // allows for text + reasoning blocks in ai messages
+ *     ai: ContentBlock.Text | ContentBlock.Reasoning;
+ *   }
+ * }
+ *
+ * // Message structure with tools and properties
+ * interface AdvancedMessageStructure extends $MessageStructure {
+ *   tools: {
+ *     calculator: $MessageToolDefinition<
+ *       { operation: string; numbers: number[] },
+ *       number
+ *     >;
+ *   };
+ *   content: {
+ *     // allows for text + image blocks in human messages
+ *     human: ContentBlock.Text | ContentBlock.Multimodal.Image;
+ *     // only allows for text blocks in ai messages
+ *     ai: ContentBlock.Text;
+ *   };
+ *   properties: {
+ *     // pins arbitrary properties to ai messages
+ *     ai: {
+ *       confidence: number;
+ *       model: string;
+ *     };
+ *   }
+ * }
+ *
+ * // Using with $MergeMessageStructure to combine structures
+ * // The resulting type when passed into BaseMessage will have a calculator tool,
+ * // allow for text + image blocks in human messages,
+ * // and text + reasoning blocks + additional arbitrary properties in ai messages.
+ * type CombinedStructure = $MergeMessageStructure<
+ *   SimpleMessageStructure,
+ *   AdvancedMessageStructure
+ * >;
+ * ```
+ */
+export interface $MessageStructure {
+  /**
+   * Optional set of tool definitions that can be used in messages.
+   * Each tool is defined with input/output types and can be referenced in tool messages.
+   */
+  readonly tools?: $MessageToolSet;
+  /**
+   * Optional mapping of message types to their allowed content blocks.
+   * Excludes the "tool" message type since tool messages have a special structure.
+   * Each message type can specify what content block types it supports (text, images, etc).
+   */
+  readonly content?: Partial<{
+    [key in $MessageType]: ContentBlock;
+  }>;
+  /**
+   * Optional mapping of message types to arbitrary property objects.
+   * Allows attaching custom metadata or other information to specific message types.
+   * Properties are typed as Record<string, unknown> to allow flexible property definitions.
```

Comment:
_nit_: can we rephrase? I think we currently only allow modifying properties that are actually present in corresponding message types, eg. this won't work:

```typescript
type S = {
  properties: {
    ai: { random: string }
  }
}

const m = new AIMessage<S>({ random: "hello" })
```